### ISSUE TITLE
Linux - Image files are overwritten with just a few images

### ISSUE THREAD

**Issue by hasidima** [wormi4ok/evernote2md#66]
Hi 

I was able to successfully convert .enex file to md files with links but the md files had the links to the same few images, although I had a lot of images in my .enex file. 
The program seems to import images 1, 2, 3 for one file, then used images for another file and named them also image1, image2, etc. effectively overwriting the previous images. Thus, I have different md file with the same images in all of them. 

Please suggest how to resolve this issue. 

Regards, 
Dmitry

**Comment by wormi4ok** [comment]
Hey Dmitry,

To work around this issue I can suggest using `--folders` flag to save every note in a separate folder with an isolated `images/` folder.

Let me know if it helps. Meanwhile, I'll try to ensure unique names for images in the next release.

**Comment by hasidima** [comment]
OK, let me try again with the folders flag

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by elzik** [comment]
Is this something that is still planned? I have the same issue and I'd really prefer not to put every note in a separate folder.

Thanks

**Comment by EnidanBri** [comment]
I do not want to draw people away from your solution, but I found [akosbalasko](https://github.com/akosbalasko)
/
[yarle](https://github.com/akosbalasko/yarle) which has the same intention, but more configuration options. 
It does rename the files both in the reference in the MD and the file itself in case the attachments have the same name.
Maybe you could join forces?

**Comment by m11751** [comment]
I want to ask a question: When I input ">evernote2md.exe --folders D:\enex_data\test.enex D:\enex_data\" , I got ever different name folder and same md name : readme.md, I want to fix the md file name with different name, how can I do or you can fix this code, thank you very much !

**Comment by wormi4ok** [comment]
Hey @m11751 ! I can offer two possible solutions:

1. The title of the file is [hardcoded here ](https://github.com/wormi4ok/evernote2md/blob/8138cbc847fd1a185ccc50a521c129174bdd5bbf/notes.go#L40)and not configurable for now. But since the code is open, you can clone this project, change one line (You would need [go installed](https://go.dev/doc/install)) and then run `go build-o evernote2md.exe` to build a your custom version)

2. A Powershell script like this should do the trick and rename all `README.md` files to `new_name.md`. Unfortunately I do not have a Windows machine to test it, so please - use it on your own risk.
```powershell
forfiles /S /M "README.md" /C "cmd /c rename @file new_name.md"
```

**Comment by StanislavMayorov** [comment]
Hi @wormi4ok. I reproduced the same bug in the latest version (0.19.0). I exported about 1000 notes successfully, but several notes have the same images. 
1-4 notes have the same image from note-4. debug logs:
```
2023/12/07 23:18:25 [DEBUG] Saving file ./temp/note-1.md
2023/12/07 23:18:25 [DEBUG] Saving attachment temp/image/Snapshot.png

2023/12/07 23:18:25 [DEBUG] Saving file ./temp/note-2.md
2023/12/07 23:18:25 [DEBUG] Saving attachment temp/image/Snapshot.png
2023/12/07 23:18:25 [DEBUG] Saving attachment temp/image/Snapshot-1.png

2023/12/07 23:18:35 [DEBUG] Saving file ./temp/note-3.md
2023/12/07 23:18:35 [DEBUG] Saving attachment temp/image/Snapshot.png

2023/12/07 23:18:37 [DEBUG] Saving file ./temp/note-4.md
2023/12/07 23:18:37 [DEBUG] Saving attachment temp/image/Snapshot.png
```

**Comment by wormi4ok** [comment]
Hi @StanislavMayorov ! Yeah, I can confirm that the issue still persists but unfortunately, it is an architectural bug in the app - the notes [are processed sequentially](https://github.com/wormi4ok/evernote2md/blob/master/main.go#L90-L120) so attachments with the same names within one note are handled correctly, but if the attachment name collides with the name in another note - it gets overwritten.
Maybe you can suggest a good way to work around this? So far I didn't manage to find time for a big refactoring that would solve the problem...

**Comment by EnidanBri** [comment]
Hi, A quick and dirty way would be to run a counter while processing, and then rename all images to have this counter in their filename. Then they would automatically be unique. 
You might have an unwanted addition to your filenames when you do this, but I think it is better than losing them. 
Basically, this is what I did when I had taht problem. But I solved it outside of your tool.

**Comment by StanislavMayorov** [comment]
Hello @wormi4ok! I think the easiest way is to add a couple of random characters at the end of file name to make all of them unique e.g. filename-d2fd86c3.pdf. Something like [useUniqueUnknownFileNames](https://github.com/akosbalasko/yarle#:~:text=the%20details%20there.-,useUniqueUnknownFileNames,-boolean)

**Comment by djol** [comment]
**evernote2md is a great tool**, but this is a *really* dangerous bug because it leads to **silent data loss**. As the size of an exported evernote notebook increases, so does the likelihood of a name collision between attachments in a note.

I unknowingly ran into this problem using evernote2md on Mac OS, exporting around 900 notes. It wasn't until I had moved the resulting md files into obsidian and used the new repository for a few weeks that I noticed the wrong images in the wrong notes. It took a little time to run down the cause of the bug as arising from the original evernote2md conversion.

Because attachments in evernote notes are often added by copy-pasting into the note or dragging a file/image, there tend to be a lot of automatic filenames created by evernote or the OS, often without any timestamp (ie. image_1, image_2....). Similarly images or attachments dragged in from the web also often have very simplistic names ("report.pdf") that a user has little control over. Name collision across these attachments are really common as the size of one's evernote database increases (in fact, the probability of attachment name collision (and thus data loss) is logarithmic in nature as the number of notes increases).

No shade on evernote2md which is a great tool, but BE VERY CAREFUL if using this to export Evernote notebooks of size containing even a moderate number of attachments. As @wormi4ok comments above, you really must use the `--folders` flag or you will be burnt by this, and you may not discover the missing attachments for a long time after completing the conversion.

**Comment by ptolemybarnes** [comment]
I would consider making the behaviour with the `--folders` flag the default in future versions, as like has been said it's easy for someone to not be aware of the data loss, perhaps until it's too late to recover from it.

**Comment by wormi4ok** [comment]
Thanks for raising this again, @ptolemybarnes . I'll try to find some time in the upcoming weeks to resolve this issue. Either following your suggestion, or fixing the underlying issue.

**Comment by wormi4ok** [comment]
I'm happy to announce that after all this time, this bug is fixed and everyone is encouraged to upgrade to [version 0.23.0](https://github.com/wormi4ok/evernote2md/releases/tag/v0.23.0)
